Adding google calendar support to a_presentation, closes issue#815

pliu2781 opened the pull request:
I added support for a_presentation helper to add to google calendar. I added a separate folder within the helpers directory that contains google_calendar_api.py, which gets called by a_presentationhelper.py. I also set it up so that the tokens retrieved by google_calendar_api.py are stored in a folder in rg-db-group/local called tokens/google_calendar. closes #815

pliu2781 commented:
For some reason, I'm getting the same Travis failure here as the one on my other PR. The test_validate tests passed locally on my machine though.

sbillinge commented on regolith/helpers/Google_Calendar_API/google_calendar_api.py:
```diff
@@ -0,0 +1,41 @@
+def add_to_calendar(event):
+    """ Takes a newly created presentation, and adds it to the user's google calendar """
+    SCOPES = ['https://www.googleapis.com/auth/calendar.events']
+    tokendir = os.path.join('tokens', 'google_calendar')
```
better to use use pathlib for this?

sbillinge commented on regolith/helpers/Google_Calendar_API/google_calendar_api.py:
```diff
@@ -0,0 +1,41 @@
+from google.oauth2.credentials import Credentials
+
+
+def add_to_calendar(event):
```
since the function is called `add_to_calendar` we probably want it to take an optional argument of `calendar_type` with a default of `"google"` and then wrap everything in `if calendar_type = "google"` 

I never did a google auth before though so I may not understand exactly what it is doing.

sbillinge commented on regolith/helpers/Google_Calendar_API/google_calendar_api.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def add_to_calendar(event):
+    """ Takes a newly created presentation, and adds it to the user's google calendar """
```
This docstring needs to be up to standard.

Also, this function looks as if it takes an _event_ and adds it to the calendar, not a presentation.  That is better tbh, because we may want to use this function in other helpers etc. and add other things to the calendar.

sbillinge commented on regolith/helpers/a_presentationhelper.py:
```diff
@@ -82,6 +83,9 @@ def subparser(subpi):
                        help="The database that will be updated.  Defaults to "
                             "first database in the regolithrc.json file.",
                        )
+    subpi.add_argument("-gc", "--google-calendar",
```
I would make this default behavior, but allow it to fail gracefully (don't crash if it fails) but add a cli arg for --no-cal in case for some reason someone wants to explicitly turn it off.

sbillinge reviewed (commented):
This is awesome!  Thanks so much.

Please see my inline comments.

Please work with someone else and see if they are getting the same travis errors and work to get to the bottom of it.  It is failing on the bad database entries, which are supposed to fail and so a fail passes the test, but does that mean that one of the databases that is supposed to fail is not failing?  weird.

pliu2781 commented:
Does the google calendar feature need a test? I'm not sure how to go about writing a test for this as this feature, as this feature needs the user to login to their google account.

sbillinge commented:
This might be hard, so no. I didn't look... How does the program behave if
someone doesn't have Google calendar or credentials?

On Fri, Jul 30, 2021, 2:10 PM pliu2781 ***@***.***> wrote:

> Does the google calendar feature need a test? I'm not sure how to go about
> writing a test for this as this feature, as this feature needs the user to
> login to their google account.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/820#issuecomment-890066512>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWULXSA5W7VQLME7U4KTT2LTJ7ANCNFSM5BDCTW6A>
> .
>

pliu2781 commented:
The program brings up a window asking the user to sign in to their google account. If they decline to grant access to regolith, then the program proceeds as normal. But if someone doesn't have a google account and closes the window, the program doesn't continue until a keyboard interrupt. I can try to fix this. Another option could also be to have the default behavior not using the google calendar.

sbillinge commented:
I see.    Let's maybe play around with it.  I guess that after the first
time you use it on a device it stores the info somewhere on your computer
and will just work, is that right?   I am not sure how easy it is, but the
best may be:


On Fri, Jul 30, 2021 at 10:32 PM pliu2781 ***@***.***> wrote:

> The program brings up a window asking the user to sign in to their google
> account. If they decline to grant access to regolith, then the program
> proceeds as normal. But if someone doesn't have a google account and closes
> the window, the program doesn't continue until a keyboard interrupt. I can
> try to fix this. Another option could also be to have the default behavior
> not using the google calendar.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/820#issuecomment-890279172>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUNMZC6UQRSPTU3QSM3T2NODXANCNFSM5BDCTW6A>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

pliu2781 commented:
> I see. Let's maybe play around with it. I guess that after the first time you use it on a device it stores the info somewhere on your computer and will just work, is that right? I am not sure how easy it is, but the best may be:
> […](#)
> On Fri, Jul 30, 2021 at 10:32 PM pliu2781 ***@***.***> wrote: The program brings up a window asking the user to sign in to their google account. If they decline to grant access to regolith, then the program proceeds as normal. But if someone doesn't have a google account and closes the window, the program doesn't continue until a keyboard interrupt. I can try to fix this. Another option could also be to have the default behavior not using the google calendar. — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub <[#820 (comment)](https://github.com/regro/regolith/pull/820#issuecomment-890279172)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAOWUNMZC6UQRSPTU3QSM3T2NODXANCNFSM5BDCTW6A> .
> -- Simon Billinge Professor, Columbia University Physicist, Brookhaven National Laboratory

After the first time, the program stores a token in a folder in rg-db-group/local called tokens. So everything works automatically after the first time. I also changed the default behavior back to not using google calendar, so the user will have to explicitly request the google calendar feature with -gc or --google-calendar.

sbillinge commented:
Ok so this is not the preferred behavior. I would like default behavior to
be Google calendar but it still works if no Google calendar. Also, I am a
little uncomfortable saving tokens into that location in case it gets
accidentally committed into the repo. It is a security risk. We usually put
tokens in the users config directory. I am not sure how ready any of that
is but that is preferred behavior

On Mon, Aug 2, 2021, 1:36 PM pliu2781 ***@***.***> wrote:

> I see. Let's maybe play around with it. I guess that after the first time
> you use it on a device it stores the info somewhere on your computer and
> will just work, is that right? I am not sure how easy it is, but the best
> may be:
> … <#m_6864982473102461884_>
> On Fri, Jul 30, 2021 at 10:32 PM pliu2781 *@*.***> wrote: The program
> brings up a window asking the user to sign in to their google account. If
> they decline to grant access to regolith, then the program proceeds as
> normal. But if someone doesn't have a google account and closes the window,
> the program doesn't continue until a keyboard interrupt. I can try to fix
> this. Another option could also be to have the default behavior not using
> the google calendar. — You are receiving this because you modified the
> open/close state. Reply to this email directly, view it on GitHub <#820
> (comment)
> <https://github.com/regro/regolith/pull/820#issuecomment-890279172>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/ABAOWUNMZC6UQRSPTU3QSM3T2NODXANCNFSM5BDCTW6A
> .
> -- Simon Billinge Professor, Columbia University Physicist, Brookhaven
> National Laboratory
>
> After the first time, the program stores a token in a folder in
> rg-db-group/local called tokens. So everything works automatically after
> the first time. I also changed the default behavior back to not using
> google calendar, so the user will have to explicitly request the google
> calendar feature with -gc or --google-calendar.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/820#issuecomment-891207798>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUK5VLDZCIE75FCM533T23JQ7ANCNFSM5BDCTW6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

sbillinge commented:
sorry, "ready" in the last sentence should have been "easy".  My phone....


On Mon, Aug 2, 2021 at 1:59 PM Simon Billinge ***@***.***> wrote:

> Ok so this is not the preferred behavior. I would like default behavior to
> be Google calendar but it still works if no Google calendar. Also, I am a
> little uncomfortable saving tokens into that location in case it gets
> accidentally committed into the repo. It is a security risk. We usually put
> tokens in the users config directory. I am not sure how ready any of that
> is but that is preferred behavior
>
> On Mon, Aug 2, 2021, 1:36 PM pliu2781 ***@***.***> wrote:
>
>> I see. Let's maybe play around with it. I guess that after the first time
>> you use it on a device it stores the info somewhere on your computer and
>> will just work, is that right? I am not sure how easy it is, but the best
>> may be:
>> … <#m_-6201887406264464911_m_6864982473102461884_>
>> On Fri, Jul 30, 2021 at 10:32 PM pliu2781 *@*.***> wrote: The program
>> brings up a window asking the user to sign in to their google account. If
>> they decline to grant access to regolith, then the program proceeds as
>> normal. But if someone doesn't have a google account and closes the window,
>> the program doesn't continue until a keyboard interrupt. I can try to fix
>> this. Another option could also be to have the default behavior not using
>> the google calendar. — You are receiving this because you modified the
>> open/close state. Reply to this email directly, view it on GitHub <#820
>> (comment)
>> <https://github.com/regro/regolith/pull/820#issuecomment-890279172>>, or
>> unsubscribe
>> https://github.com/notifications/unsubscribe-auth/ABAOWUNMZC6UQRSPTU3QSM3T2NODXANCNFSM5BDCTW6A
>> .
>> -- Simon Billinge Professor, Columbia University Physicist, Brookhaven
>> National Laboratory
>>
>> After the first time, the program stores a token in a folder in
>> rg-db-group/local called tokens. So everything works automatically after
>> the first time. I also changed the default behavior back to not using
>> google calendar, so the user will have to explicitly request the google
>> calendar feature with -gc or --google-calendar.
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/regro/regolith/pull/820#issuecomment-891207798>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABAOWUK5VLDZCIE75FCM533T23JQ7ANCNFSM5BDCTW6A>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
>> .
>>
>

-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

pliu2781 commented:
I changed the google calendar functionality back to default. I also fixed the issue where the program will wait for the google calendar request to finish before dumping. The program still waits for a response from the user allowing access to regolith, so non google users will either have to keyboard interrupt or toggle --no-cal.

sbillinge commented on regolith/commands.py:
```diff
@@ -247,3 +248,11 @@ def validate(rc):
     else:
         sys.exit("Validation failed on some records")
+
+def google_calendar_updater(rc):
```
this should be in tools.py, not in commands.py.  It is not a regolith command.

sbillinge commented on regolith/helpers/Google_Calendar_API/credentials.json:
```diff
@@ -0,0 +1 @@
+{"installed":{"client_id":"154057909362-e30si590e8n6n6p4lljth074kj7dgv9o.apps.googleusercontent.com","project_id":"regolith-addprojectum-calendar","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"gmxkVOmAn5tY8SVJTMswtIpG","redirect_uris":["urn:ietf:wg ...
```
what is this?  I am guessing it shouldn't be here.

It will live forever in the git history, so if this is a valid token it should be changed on the Google side so it doesn't give access anywhere.

pliu2781 replied:
This is a credentials file that contains client_id and client_secret that allows people to access the api. https://cloud.google.com/docs/authentication/end-user The documentation here says this json file has to be distributed with the application. This other documentation https://developers.google.com/identity/protocols/oauth2 says: 'The process results in a client ID and, in some cases, a client secret, which you embed in the source code of your application. (In this context, the client secret is obviously not treated as a secret.)' However, I'm not an expert with this, so distributing this info does seem kind of insecure to me.

sbillinge commented on regolith/helpers/Google_Calendar_API/google_calendar_api.py:
```diff
@@ -0,0 +1,47 @@
+from __future__ import print_function
```
I think this stuff should also be in tools and no Google_Calendar_API (and therefore no modules in there), so as not to change the API of regolith for some small side-case of Google calendar functionality.

sbillinge commented on regolith/helpers/a_presentationhelper.py:
```diff
@@ -82,6 +84,9 @@ def subparser(subpi):
                        help="The database that will be updated.  Defaults to "
                             "first database in the regolithrc.json file.",
                        )
+    subpi.add_argument("-nc", "--no_cal",
```
this won't work, the shortcut has to be a single letter.   Just delete `-nc` so the user has to type the long name.

sbillinge commented on regolith/main.py:
```diff
@@ -349,6 +349,7 @@ def main(args=None):
             dbs = commands.helper_db_check(rc)
         with connect(rc, dbs=dbs) as rc.client:
             CONNECTED_COMMANDS[rc.cmd](rc)
+        commands.google_calendar_updater(rc)
```
I am pretty sure this is the wrong pattern, but we may need to discuss it if there is some reason it has to be this way.

pliu2781 replied:
I put the google calendar function in here, because I wanted it to be outside of the context manager. That way, the program will dump the presentation files first without having to wait for the google calendar user consent. Is it better to have this function in a_presentationhelper.py?

sbillinge replied:
This is a pretty major perturbation of the whole structure of the Regolith package, just as a workaround for a small annoyance.  It breaks all the logic will be hard to maintain in the future because we won't know why it is there breaking the logic, so I dislike it very much, though I understand your motivation.   It probably needs more discussion and a better solution.  I am guessing that there is a better solution than this if we can put our thinking caps on .

In general, we want our code to structured to be understandable and maintainable and there has to be a VERY good reason to do something that breaks logic.

sbillinge replied:
can you discuss it maybe with Zach and Connor?

pliu2781 replied:
Sure, I'll reach out to them

sbillinge reviewed (commented):
please see comments.  We may have to have an in person meeting to resolve some of these things if I am not understanding the issues well enough.

sbillinge commented:
ok.  I am not expert either.  Can we have Zach look at it.  We can maybe
look at some other packages that do this and see how they do it.

On Mon, Aug 2, 2021 at 8:12 PM pliu2781 ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In regolith/helpers/Google_Calendar_API/credentials.json
> <https://github.com/regro/regolith/pull/820#discussion_r681343795>:
>
> > @@ -0,0 +1 @@
> +{"installed":{"client_id":"154057909362-e30si590e8n6n6p4lljth074kj7dgv9o.apps.googleusercontent.com","project_id":"regolith-addprojectum-calendar","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"gmxkVOmAn5tY8SVJTMswtIpG","redirect_uris":["urn:ietf:wg:oauth:2.0:oob","http://localhost"]}}
>
> This is a credentials file that contains client_id and client_secret that
> allows people to access the api.
> https://cloud.google.com/docs/authentication/end-user The documentation
> here says this json file has to be distributed with the application. This
> other documentation
> https://developers.google.com/identity/protocols/oauth2 says: 'The
> process results in a client ID and, in some cases, a client secret, which
> you embed in the source code of your application. (In this context, the
> client secret is obviously not treated as a secret.)' However, I'm not an
> expert with this, so distributing this info does seem kind of insecure to
> me.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/820#discussion_r681343795>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOIGB6ONDLJQKLGSMTT24X6ZANCNFSM5BDCTW6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

pliu2781 commented:
> ok. I am not expert either. Can we have Zach look at it. We can maybe look at some other packages that do this and see how they do it.
> […](#)
> On Mon, Aug 2, 2021 at 8:12 PM pliu2781 ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In regolith/helpers/Google_Calendar_API/credentials.json <[#820 (comment)](https://github.com/regro/regolith/pull/820#discussion_r681343795)>: > @@ -0,0 +1 @@ +{"installed":{"client_id":"154057909362-e30si590e8n6n6p4lljth074kj7dgv9o.apps.googleusercontent.com","project_id":"regolith-addprojectum-calendar","auth_uri":"[https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"gmxkVOmAn5tY8SVJTMswtIpG","redirect_uris":["urn:ietf:wg:oauth:2.0:oob","http://localhost"]](https://accounts.google.com/o/oauth2/auth%22,%22token_uri%22:%22https://oauth2.googleapis.com/token%22,%22auth_provider_x509_cert_url%22:%22https://www.googleapis.com/oauth2/v1/certs%22,%22client_secret%22:%22gmxkVOmAn5tY8SVJTMswtIpG%22,%22redirect_uris%22:%5B%22urn:ietf:wg:oauth:2.0:oob%22,%22http://localhost%22%5D)}} This is a credentials file that contains client_id and client_secret that allows people to access the api. https://cloud.google.com/docs/authentication/end-user The documentation here says this json file has to be distributed with the application. This other documentation https://developers.google.com/identity/protocols/oauth2 says: 'The process results in a client ID and, in some cases, a client secret, which you embed in the source code of your application. (In this context, the client secret is obviously not treated as a secret.)' However, I'm not an expert with this, so distributing this info does seem kind of insecure to me. — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub <[#820 (comment)](https://github.com/regro/regolith/pull/820#discussion_r681343795)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAOWUOIGB6ONDLJQKLGSMTT24X6ZANCNFSM5BDCTW6A> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email> .
> -- Simon Billinge Professor, Columbia University Physicist, Brookhaven National Laboratory

Ill reach out to him

sbillinge commented on regolith/tools.py:
```diff
@@ -2039,8 +2039,8 @@ def add_to_google_calendar(event):
         else:
-            print('If you want to enable the google calendar feature, please run the helper with the flag -f or --flow'
-                  'This only has to be done once.')
+            print('If you want to activate the google calendar feature, please rerun the helper with the flag'
```
don't ask what they want, but tell what they do.  How about:
"The google calendar feature needs authentication information to run.  This needs to be done just once for each new device. To do this, please type `regolith helper a_presentation --activate-google-calendar` and follow the prompts in the browser popup from Google.  You can then rerun this command and it should work, now and in the future.\n\n The helper may be run without creating a calendar event by rerunning the command (e.g., use up-arrow) and adding --no-cal."

sbillinge reviewed (commented):
one last tweak.  Very much looking forward to this...

pliu2781 commented:
Travis is saying the build errors because of `ModuleNotFoundError: No module named 'googleapiclient'`

sbillinge commented:
> 
> 
> Travis is saying the build errors because of `ModuleNotFoundError: No module named 'googleapiclient'`

did you add this requirement to requirements.txt?

pliu2781 commented:
> > Travis is saying the build errors because of `ModuleNotFoundError: No module named 'googleapiclient'`
> 
> did you add this requirement to requirements.txt?

I just added the google modules to requirements/run.txt. Lets see if it works now

sbillinge commented:
I am excited to test this....thanks Peter.  I added a new issue to make tests for the google api, but that can wait for the future.  It will involve mocking the api call.  For sure we will need to know how to do this in the future for multiple cases, so that can be the job of a future student.....

pliu2781 commented:
One more thing:
Should replace the credentials.json file with one from the group's google cloud account instead of my private one?

sbillinge commented:
yes.  What is this credentials for?

On Tue, Aug 17, 2021 at 9:57 AM pliu2781 ***@***.***> wrote:

> One more thing:
> Should replace the credentials.json file with one from the group's google
> cloud account instead of my private one?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/820#issuecomment-900321825>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOXNMFHPKAI234HI53T5JTF3ANCNFSM5BDCTW6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

pliu2781 commented:
The credentials.json is generated by the google cloud account. It contains a client_id and client_secret that lets google keep track of how many api calls the program is making.

sbillinge commented:
Ok. Please send step by step instructions for me to make the switch

On Tue, Aug 17, 2021, 10:22 AM pliu2781 ***@***.***> wrote:

> The credentials.json is generated by the google cloud account. It contains
> a client_id and client_secret that lets google keep track of how many api
> calls the program is making.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/820#issuecomment-900342454>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUKQJB2FLY5FWQ6H4JLT5JWA7ANCNFSM5BDCTW6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

pliu2781 commented:
https://developers.google.com/workspace/guides/create-credentials
